src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1302 @@
+            Ok(())
+        }
+    }
+
+    fn read_node(&mut self, ptr: &TriePtr) -> Result<(TrieNodeType, TrieHash), Error> {
+        let disk_ptr = ftell(self)?;
+        trace!("TrieRAM: read_node({:?}): at {}: {:?}", &self.block_header, disk_ptr, ptr);
+
+        self.read_count += 1;
+        if is_backptr(ptr.id()) {
+            self.read_backptr_count += 1;
+        }
+        else if ptr.id() == TrieNodeID::Leaf {
+            self.read_leaf_count += 1;
+        }
+        else {
+            self.read_node_count += 1;
+        }
+
+        if (ptr.ptr() as u64) >= (self.data.len() as u64) {
+            trace!("TrieRAM: Failed to read node: {} >= {}", ptr.ptr(), self.data.len());
+            Err(Error::NotFoundError)
+        }
+        else {
+            Ok(self.data[ptr.ptr() as usize].clone())
+        }
+    }
+
+    fn write_node(&mut self, node: &TrieNodeType, hash: TrieHash) -> Result<(), Error> {
+        if self.readonly {
+            trace!("Read-only!");
+            return Err(Error::ReadOnlyError);
+        }
+
+        let disk_ptr = ftell(self)?;
+        trace!("TrieRAM: write_node({:?}): at {}: {:?} {:?}", &self.block_header, disk_ptr, &hash, node);
+        
+        self.write_count += 1;
+        match node {
+            TrieNodeType::Leaf(_) => {
+                self.write_leaf_count += 1;
+            },
+            _ => {
+                self.write_node_count += 1;
+            }
+        }
+
+        if self.offset < (self.data.len() as u64) {
+            self.data[self.offset as usize] = (node.clone(), hash);
+            self.offset += 1;
+            Ok(())
+        }
+        else if self.offset == (self.data.len() as u64) {
+            self.data.push((node.clone(), hash));
+            self.offset += 1;
+            self.num_nodes += 1;
+            self.total_bytes += get_node_byte_len(node);
+            Ok(())
+        }
+        else {
+            trace!("Failed to write node bytes: off the end of the buffer");
+            Err(Error::NotFoundError)
+        }
+    }
+
+    fn flush(&mut self) -> Result<(), Error> {
+        Ok(())
+    }
+
+    fn num_blocks(&self) -> usize {
+        1
+    }
+
+    fn chain_tips(&self) -> Vec<BlockHeaderHash> {
+        vec![self.block_header.clone()]
+    }
+}
+
+impl Seek for TrieRAM {
+    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+        match pos {
+            SeekFrom::Start(ref loc) => {
+                let prev = self.offset;
+                self.offset = *loc;
+                Ok(prev)
+            },
+            SeekFrom::End(ref loc) => {
+                let prev = self.num_nodes;
+                let abs_loc = (*loc).abs() as u64;
+                if abs_loc > self.num_nodes {
+                    // can't seek behind 0
+                    return Err(io::Error::new(io::ErrorKind::InvalidInput, Error::BadSeekValue));
+                }
+
+                let new_offset = (self.num_nodes as i128) + (*loc as i128);
```

Comment:
I think you can avoid casting and performin i128 arithmetic here (and the manual overflow check) by using checked_add:

```
self.offset = self.num_nodes.checked_add(loc)
  .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, Error::BadSeekValue))?;
```

Reply:
This should be addressed now.